### PR TITLE
Replace TokenizedRegularExpression with PHPParser

### PR DESCRIPTION
**Pull request by dhensby**
see #5549

This is a PoC that we can remove our home baked static tokenizer for a library.

This reduces our code maintenance but at the cost of higher memory and cpu usage... which needs addressing

~~Depends on #5572 (needs merging first)~~

### CODE REVIEW

**Comment by dhensby** [comment]
So @kinglozzer has compared this PR with current master and there's a fairly significant impact on time and memory:

image

**Comment by kinglozzer** [comment]
Looks like there could be some memory overhead added by XHProf... using `memory_get_peak_usage(true)` reports 22mb for current master and 25mb for this PR. Still an increase, but around half of what XHProf was reporting.

Using `memory_get_peak_usage(true)` and `time` on CLI reports:

```
Current master:
22mb
real 0m2.620s
user 0m2.463s
sys 0m0.139s

This PR:
25.5mb
real 0m5.068s
user 0m4.843s
sys 0m0.193s
```

**Comment by SpiritLevel** [comment]
@dhensby, @kinglozzer  Dan, Loz, is that a Github tool you are using to compare runs ? Where can I find it ?

**Comment by dhensby** [comment]
No, we are running locally with CLI

**Comment by dhensby** [comment]
OK, here is my bench marking:

```
Test results - tested on XS cloud server by atlantic.net

classmanifest branch:
100 runs of building brand new manifest	: 20m17.015s
100 runs of flushing manifest: 0m42.927s

current master:
100 runs of building brand new manifest	: 7m50.568s
100 runs of flushing manifest: 0m43.211s

diff:
Brand new: 12m26.447s = 7.46s per run
Flushing: -0.284s = negligible

Memory usage: (rm -rf silverstripe-cache/* && framwork/sake /)
classmanifest: 25.75 mb (27000832)
master: 22.25 mb (23330816)
diff: 3.5 mb (3670016)
```

We really need to find a way to optimise this because a 7.5s increase on building the manifest from scratch is not acceptable

**Comment by dhensby** [comment]
Related: https://github.com/nikic/PHP-Parser/issues/278

**Comment by sminnee** [comment]
I'd be surprised if 278 gets results, but it's worth a shot.

In the meantime let's stick with TRE and making minor amendments to that.

**Comment by tractorcow** [comment]
We also have a lot of logic for tokenisation in Object::parse_class_spec. I'd like to, at the very least, see some kind of consolidation for this logic so we don't have custom baked parsers sitting around the code base.

**Comment by sminnee** [comment]
> We also have a lot of logic for tokenisation in Object::parse_class_spec. I'd like to, at the very least, see some kind of consolidation for this logic so we don't have custom baked parsers sitting around the code base.

It's quite a different parsing challenge, although it would probably benefit from using the built-in tokeniser.

**Comment by sminnee** [comment]
@dhensby Given that `php-parser` said that your suggested change wasn't possible; do you think that we're likely to get this to be faster than the current implementation?

**Comment by dhensby** [comment]
OK - so #6615 (and my solution #6619) go to show how hand-rolling our own file parser can cause us maintenance problems. I've had to fudge my fix a little because I can't get it to work the way I'd like (removing references to `T_WHITESPACE` from the TRE definitions - I've had to just mark them as optional for now).

Whilst a 7s increase on initial build of the manifest is a lot, perhaps it's something that isn't sooooo bad if we consider it's only on a completely fresh manifest build (which is quite rare)?

The only other way I can think of getting this faster is working on making https://github.com/nikic/PHP-Parser faster by having it only instantiate objects when entering a syntax tree rather than instantiating the entire AST for a file and then performing the search over it. This is based on the assumption the the extra compute time comes from instantiating the token objects.

If we're only going to enter 1 of the syntax trees, then it's not worth the CPU cycles to instantiate all of them... I'd imagine this is quite a bit of work and would be a 3.x release of PHPParser, so not something that could happen any time soon...

**Comment by dhensby** [comment]
In fact, one of @kinglozzer's original ideas was to only send the tokens through to PHPParser that we care about. so strip out the vast majority of tokens.

And, another point, we don't have 7.x PHP syntax support in our parser, something like this would not be recognised in SS: `use Microsoft\PhpParser\{DiagnosticsProvider, Node, Parser, PositionUtilities};`

**Comment by dhensby** [comment]
OK - with this fix I'm seeing performance improvements over current master of ~1second

**Comment by dhensby** [comment]
so we need to test the php7 style of import for some confidence in this.

File: src/Core/Manifest/NodeVisitor.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace SilverStripe\Core\Manifest;
+
+use PhpParser\Node;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitorAbstract;
+
+class NodeVisitor extends NodeVisitorAbstract
```
Comment (sminnee) [inline src/Core/Manifest/NodeVisitor.php]: You can pick a better name than this.

**Comment by tractorcow** [comment]
I think cleanFileContents needs some tests of its own, since it's custom code.

File: src/Core/Manifest/ClassManifest.php [hunk 2]
```diff
@@ -644,36 +418,89 @@ public function handleFile($basename, $pathname, $depth)
...
+    }
+
+    /**
+     * Clean out any information from PHP files that is nested more than $cutOffDepth in curly braces
+     *
+     * @param string $contents
+     * @param int $cutOffDepth
+     * @return string
+     */
+    protected function cleanFileContents($contents, $cutOffDepth = 1)
```
Comment (sminnee) [inline src/Core/Manifest/ClassManifest.php]: Add a test specifically for cleanFileContents. Related to that, maybe burying it as a protected method of this class isn't as good as just splitting it out into a separate class.

Maybe `ClassContentRemover` is a better name?

File: src/Core/Manifest/NodeVisitor.php (hunk 1 above)
Comment (dhensby) [inline src/Core/Manifest/NodeVisitor.php]: How about `ManifestVistor`? I was thinking the NS provides enough context here (`SilverStripe\Core\Manifest\NodeVisitor`). I could rename it during import `use SilverStripe\Core\Manifest\NodeVisitor as ManifestNodeVisitor` for more clarity outside the class?

Any other suggestions welcome

**Comment by dhensby** [comment]
Performance test results:

Current master (449ed8d379): 5m37.053s (3.37s avg)
This code: 3min 14 sec (1.94s avg)

~40% reduction in execution time

---

Test script:

```
#!/bin/bash

for i in {1..100}; do
        rm -rf silverstripe-cache/*
        framework/sake / > /dev/null
done
```

**Comment by sminnee** [comment]
Regarding name: maybe ClassManifestVisitor?

**Comment by dhensby** [comment]
OK - I think this is ready now.

I've re run the tests against latest master and can see a speed improvement of 400ms on average for this vs latest master.

File: src/Core/Manifest/ClassManifest.php [hunk 3]
```diff
@@ -195,6 +93,35 @@ public function __construct($base, $includeTests = false, $forceRegen = false, $
         }
     }
 
+    public function getParser()
+    {
+        if (!$this->parser) {
+            $this->parser = (new ParserFactory)->create(ParserFactory::PREFER_PHP5);
```
Comment (kinglozzer) [inline src/Core/Manifest/ClassManifest.php]: Any reason for `PREFER_PHP5` over 7? [PHP-Parser docs](https://github.com/nikic/PHP-Parser/blob/master/doc/2_Usage_of_basic_components.markdown#parsing) say:

> Unless you have strong reason to use something else, `PREFER_PHP7` is a reasonable default.

The PHP 7 grammar looks to have slightly fewer rules, is it possibly marginally more performant that way around?

File: src/Core/Manifest/ClassManifest.php [hunk 4]
```diff
@@ -662,40 +435,42 @@ public function handleFile($basename, $pathname, $depth)
...
-        foreach ($traits as $trait) {
-            $this->traits[strtolower($namespaceBase . $trait['traitName'])] = $pathname;
+        foreach ($traits as $traitName => $traitInfo) {
+            $this->traits[strtolower($traitName)] = $pathname;
         }
     }
 
     /**
-     * Recursively coalesces direct child information into full descendant
+     * Recursively coalthat esces direct child information into full descendant
```
Comment (kinglozzer) [inline src/Core/Manifest/ClassManifest.php]: Typo

**Comment by dhensby** [comment]
OK - ready ready. we are now seeing average improvement on building empty caches of... 1.45 seconds vs current master

**Comment by kinglozzer** [comment]
Here’s an XHProf comparison of current master vs this PR:

screen shot 2017-03-02 at 16 52 11

I don’t know if/how much these results are being skewed by https://github.com/silverstripe/silverstripe-framework/issues/6670

**Comment by dhensby** [comment]
Ping @sminnee

**Comment by tractorcow** [comment]
How well does this code with code such as:

```php
class SomeClass {
	function bob() {
		return '{' . '{' . '{';
	}
}

class AnotherClass {
}
```

**Comment by dhensby** [comment]
It works great, test added as proof.

**Comment by dhensby** [comment]
we have lift off! 🚀

**Pull request merged**